client/post-editor/editor-term-selector/index.jsx

```diff
@@ -26,21 +30,34 @@ class EditorTermSelector extends Component {
 	}
 
 	onTermsChange( selectedTerm ) {
-		const { postTerms, taxonomyName } = this.props;
+		const { postTerms, taxonomyName, siteId, postId } = this.props;
 		const terms = cloneDeep( postTerms ) || {};
-		const taxonomyTerms = terms[ taxonomyName ] || {};
+		let taxonomyTerms = map( terms[ taxonomyName ] );
+		const termIds = map( taxonomyTerms, 'ID' );
+
+		if ( includes( termIds, selectedTerm.ID ) ) {
+			taxonomyTerms = taxonomyTerms.reduce( ( prev, term ) => {
+				if ( term.ID === selectedTerm.ID ) {
+					return prev;
+				}
 
-		if ( taxonomyTerms[ selectedTerm.name ] ) {
-			delete taxonomyTerms[ selectedTerm.name ];
+				prev.push( term );
+				return prev;
+			}, [] );
 		} else {
-			taxonomyTerms[ selectedTerm.name ] = selectedTerm;
+			taxonomyTerms.push( selectedTerm );
 		}
-		terms[ taxonomyName ] = taxonomyTerms;
 
-		// TODO: REDUX - remove flux actions when whole post-editor is reduxified
-		PostActions.edit( {
-			terms: terms
-		} );
+		const taxonomyTermIds = map( taxonomyTerms, 'ID' );
+
+		this.props.editPost( {
+			terms: {
+				[ taxonomyName ]: taxonomyTerms
+			},
+			terms_by_id: {
+				[ taxonomyName ]: taxonomyTermIds.length ? taxonomyTermIds : null
+			}
+		}, siteId, postId );
```

Comment:
>  So that they're kept in sync in the getEditedPost return value

Correct.  This was being done prior by the `normalizer()` in the `post-edit-store`.

> Maybe the return value of getEditedPost could be normalized such that we don't include the full terms object, only terms_by_id

This is an option - but having the full object is currently used by the `<TermTokenField />` and am guessing will also be used to populate the selected labels in the taxonomy accordions.

We have had to do quite a bit of hackery to get terms to work properly here so I think the best long-term option is to strive for using the full `terms` object throughout, but change the shape of it to:

```
terms: {
    taxonomyName: [ { term }, { anotherTerm } ]
}
```

This will fix the problem with duplicate hierarchical term names, and the issue @nylen dealt with in `merge` yesterday.

The reason for leaning on `terms_by_id` here is the same as it was in #6468 - again the API not being able to handle the updating of `terms` if it is sent the same shape of an object that it responds with.